mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,328 @@
+        for the fluid along with a vector of species masses for multi-component
+        fluids.
+
+    boundaries: float
+
+        Dictionary of boundary functions, one for each valid boundary tag
+
+    t: float
+
+        Time
+
+    alpha: float
+
+       The maximum artificial viscosity coefficient to be applied
+
+    eos: :class:`~mirgecom.eos.GasEOS`
+
+       Only used as a pass through to the boundary conditions.
+
+    Returns
+    -------
+    numpy.ndarray
+
+        The artificial viscosity operator applied to *q*.
+    """
+    # Get smoothness indicator based on first component
+    indicator_field = q[0] if isinstance(q, np.ndarray) else q
+    indicator = smoothness_indicator(discr, indicator_field, **kwargs)
+
+    # R=Grad(Q) volume part
+    if isinstance(q, np.ndarray):
+        grad_q_vol = np.stack(obj_array_vectorize(discr.weak_grad, q), axis=0)
+    else:
+        grad_q_vol = discr.weak_grad(q)
+
+    # R=Grad(Q) Q flux over interior faces
+    q_flux_int = _facial_flux_q(discr, q_tpair=interior_trace_pair(discr, q))
+    # R=Grad(Q) Q flux interior faces on partition boundaries
+    q_flux_pb = sum(_facial_flux_q(discr, q_tpair=pb_tpair)
+                    for pb_tpair in cross_rank_trace_pairs(discr, q))
+    # R=Grad(Q) Q flux domain boundary part (i.e. BCs)
+    q_flux_db = 0
+    for btag in boundaries:
+        q_tpair = TracePair(
+            btag,
+            interior=discr.project("vol", btag, q),
+            exterior=boundaries[btag].exterior_q(discr, btag=btag, t=t,
+                                                 q=q, eos=eos))
+        q_flux_db = q_flux_db + _facial_flux_q(discr, q_tpair=q_tpair)
+    # Total Q flux across element boundaries
+    q_bnd_flux = q_flux_int + q_flux_pb + q_flux_db
+
+    # Compute R
+    r = discr.inverse_mass(
+        -alpha * indicator * (grad_q_vol - discr.face_mass(q_bnd_flux))
+    )
+
+    # RHS_av = div(R) volume part
+    div_r_vol = discr.weak_div(r)
+    # RHS_av = div(R): grad(Q) flux interior faces part
+    r_flux_int = _facial_flux_r(discr, r_tpair=interior_trace_pair(discr, r))
+    # RHS_av = div(R): grad(Q) flux interior faces on the partition boundaries
+    r_flux_pb = sum(_facial_flux_r(discr, r_tpair=pb_tpair)
+                    for pb_tpair in cross_rank_trace_pairs(discr, r))
+    # RHS_av = div(R): grad(Q) flux domain boundary part (BCs)
+    r_flux_db = 0
+    for btag in boundaries:
+        r_tpair = TracePair(
+            btag,
+            interior=discr.project("vol", btag, r),
+            exterior=boundaries[btag].exterior_grad_q(discr, btag=btag, t=t,
+                                                      grad_q=r, eos=eos))
+        r_flux_db = r_flux_db + _facial_flux_r(discr, r_tpair=r_tpair)
+    # Total grad(Q) flux element boundaries
+    r_flux_bnd = r_flux_int + r_flux_pb + r_flux_db
+
+    # Return the AV RHS term
+    return discr.inverse_mass(-div_r_vol + discr.face_mass(r_flux_bnd))
```

Comment:
w-hagen/mirgecom#21

I had to refactor the diffusion operator somewhat so that it could take in fluid BC objects. @MTCam tells me the BC APIs are currently in flux (heh), so this may need further tweaking to be compatible with that.

Otherwise it seems like it works... The AV tests pass, and the doublemach-mpi case looks visually similar (though not numerically identical) after 100 steps. I suspect the differences come down to the AV implementation and the diffusion operator using different conventions for the auxiliary variable (`alpha*grad(u)` vs. `grad(u)`). Are there any more tests we can run to make sure this is ok @w-hagen?

Reply:
Great. Thanks Matt, I have a couple of ideas rolling around in my head at the moment on how we could possibly add some more testing to this. Let me think on it a bit more and maybe try to implement some additional tests.